core/src/main/java/com/graphhopper/routing/weighting/custom/CustomWeightingHelper.java

```diff
@@ -0,0 +1,67 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.weighting.custom;
+
+import com.graphhopper.json.geo.JsonFeature;
+import com.graphhopper.routing.ev.DecimalEncodedValue;
+import com.graphhopper.routing.ev.EncodedValueLookup;
+import com.graphhopper.util.EdgeIteratorState;
+import com.graphhopper.util.FetchMode;
+import com.graphhopper.util.GHUtility;
+import com.graphhopper.util.shapes.BBox;
+import com.graphhopper.util.shapes.Polygon;
+
+import java.util.Map;
+
+/**
+ * This class is for internal usage only. It is subclassed by Janino, then special expressions are injected into init,
+ * getSpeed and getPriority. At the end an instance is created and used in CustomWeighting.
+ */
+public class CustomWeightingHelper {
+    protected DecimalEncodedValue avg_speed_enc;
+    protected static boolean DEFAULT = true;
+
+    protected CustomWeightingHelper() {
+    }
+
+    public void init(EncodedValueLookup lookup, DecimalEncodedValue avgSpeedEnc, Map<String, JsonFeature> areas) {
+        this.avg_speed_enc = avgSpeedEnc;
+    }
+
+    public double getPriority(EdgeIteratorState edge, boolean reverse) {
+        return 1;
+    }
+
+    public double getSpeed(EdgeIteratorState edge, boolean reverse) {
+        return getRawSpeed(edge, reverse);
+    }
+
+    protected final double getRawSpeed(EdgeIteratorState edge, boolean reverse) {
+        double speed = reverse ? edge.getReverse(avg_speed_enc) : edge.get(avg_speed_enc);
+        if (Double.isInfinite(speed) || Double.isNaN(speed) || speed < 0)
+            throw new IllegalStateException("Invalid estimated speed " + speed);
+        return speed;
+    }
+
+    public static boolean in(Polygon p, EdgeIteratorState edge) {
+        BBox bbox = GHUtility.createBBox(edge);
+        if (p.getBounds().intersects(bbox))
+            return p.intersects(edge.fetchWayGeometry(FetchMode.ALL).makeImmutable()); // TODO PERF: cache bbox and edge wayGeometry for multiple area
+        return false;
+    }
```

Comment:
We discussed where to place this method in another comment, but it seems this is not even used?

Reply:
It is used inside the generated script when areas are involved. And the script has access to this method because it is a subclass. The "script" is currently called Test -> will rename it to CustomWeightingHelperSubclass?

Reply:
Where? I tried finding this via text search but could not find it. 

> `CustomWeightingHelperSubclass`

I like this better than `Test` but it still doesn't say much. Maybe include `Janino`?

Reply:
Ok

> Where? I tried finding this via text search but could not find it.

See e.g. CustomWeightingTest.testArea there we use in_area_custom1 and such a variable is replaced with the method call via `ExpressionVisitor.injects`. See also https://github.com/graphhopper/graphhopper/pull/2209/commits/dc4268a4b883760e4b727999ee98725a02244b53 where I moved from:

```java
getSpeed(...) {
  boolean in_area_custom1 = areas.get("custom1"); // area_custom1 is initialized in the init method via areas.get("custom1")
  double speed = /*getSpeed from edge*/;

  // now in_area_custom1 can be used as boolean variable
  if (some_expression && in_area_custom1) 
     speed *= 0.5;

  return speed;
}
```

to 

```java
getSpeed(...) {
  double speed = /*getSpeed from edge*/;

  // call in method directly so we call it only if some_expression == true
  if (some_expression && in(in_area_custom1, edge)) 
    speed *= 0.5;

  return speed;
}
```